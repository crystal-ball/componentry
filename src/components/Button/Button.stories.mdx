import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs'
import { Button } from './Button'

<Meta title='Components/Button' component={Button} />

export const Template = (args) => <Button {...args} />

# Button

## Variants

Pass a `filled` or `outlined` variant to set a button style.

<Story
  name='Button'
  decorators={[
    (Story) => (
      <div className='grid grid-cols-4 gap-x-4 gap-y-2 items-center justify-items-center w-1/2'>
        <div>base</div>
        <div>hover</div>
        <div>active</div>
        <div>disabled</div>
        <Story />
      </div>
    ),
  ]}
>
  <Button variant='filled'>Button</Button>
  <Button variant='filled' className='C9Y-hover'>
    Button
  </Button>
  <Button variant='filled' className='C9Y-active'>
    Button
  </Button>
  <Button variant='filled' disabled>
    Button
  </Button>
  <Button variant='outlined'>Button</Button>
  <Button variant='outlined' className='C9Y-hover'>
    Button
  </Button>
  <Button variant='outlined' className='C9Y-active'>
    Button
  </Button>
  <Button variant='outlined' disabled>
    Button
  </Button>
</Story>

## Sizes

Pass a `size` to set a button size:

<div className='flex flex-row items-center gap-4'>
  <Button size='small'>Button</Button>
  <Button>Button</Button>
  <Button size='large'>Button</Button>
</div>

## Full width buttons

Pass `fullWidth` to make a button fill its parent container:

<Button fullWidth>Button</Button>

## Disabled

- Controlled by `disabled` prop
- Adds `disabled` attribute to element and `.C9Y-Disabled` class
- Adds `pointer-events: none` to prevent user actions
- Prop `wrapWhenDisabled` controls wrapping disabled buttons with a span to support
  accessing a ref (eg for tooltip) and to support setting `cursor: not-allowed` which has
  no effect on `pointer-events: none` button element

<div className='flex flex-col items-center gap-4'>
  <div className='flex flex-row items-center gap-4'>
    <Button variant='filled' disabled>
      Button
    </Button>
    <Button variant='outlined' disabled>
      Button
    </Button>
  </div>
  <Button disabled fullWidth>
    Button
  </Button>
</div>

## Icons

- Button supports `startIcon` and `endIcon` props
- Icon sizes are defined with a single level selector (important for ability to easily
  override by consumer)
- Button spacing is defaulted with `gap`

<div className='flex flex-col items-center gap-4'>
  <div className='flex gap-4'>
    <Button variant='filled' startIcon='code' size='small'>
      Button
    </Button>
    <Button variant='filled' endIcon='code' size='small'>
      Button
    </Button>
  </div>
  <div className='flex gap-4'>
    <Button variant='filled' startIcon='code'>
      Button
    </Button>
    <Button variant='filled' endIcon='code'>
      Button
    </Button>
  </div>
  <div className='flex gap-4'>
    <Button variant='filled' startIcon='code' size='large'>
      Button
    </Button>
    <Button variant='filled' endIcon='code' size='large'>
      Button
    </Button>
  </div>
</div>

## Buttons with hrefs

Buttons with an `href` will be rendered as anchors:

<div className='flex flex-col items-center gap-4'>
  <Button href='#test'>Button</Button>
  <Button disabled href='#test'>
    Button
  </Button>
</div>

<ArgsTable of={Button} />
