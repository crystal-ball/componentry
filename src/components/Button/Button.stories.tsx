import type { Meta, StoryObj } from '@storybook/react'

import { Button } from './Button'

const meta: Meta<typeof Button> = {
  component: Button,
}

export default meta
type Story = StoryObj<typeof Button>

export const Primary: Story = {
  decorators: [
    (Story) => (
      <div className='grid grid-cols-4 gap-x-4 gap-y-2 items-center justify-items-center'>
        <div>base</div>
        <div>hover</div>
        <div>active</div>
        <div>disabled</div>
        <Story />
      </div>
    ),
  ],
  render: () => (
    <>
      <Button variant='filled'>Button</Button>
      <Button variant='filled' className='C9Y-hover'>
        Button
      </Button>
      <Button variant='filled' className='C9Y-active'>
        Button
      </Button>
      <Button variant='filled' disabled>
        Button
      </Button>
      <Button variant='outlined'>Button</Button>
      <Button variant='outlined' className='C9Y-hover'>
        Button
      </Button>
      <Button variant='outlined' className='C9Y-active'>
        Button
      </Button>
      <Button variant='outlined' disabled>
        Button
      </Button>
    </>
  ),
}

export const WithIcon: Story = {
  render: () => (
    <div className='flex flex-row items-center gap-4'>
      <Button variant='filled' size='small' startIcon='code' endIcon='code'>
        Button
      </Button>
      <Button variant='filled' startIcon='code' endIcon='code'>
        Button
      </Button>
      <Button variant='filled' size='large' startIcon='code' endIcon='code'>
        Button
      </Button>
    </div>
  ),
}

export const Sizes: Story = {
  render: () => (
    <div className='flex flex-row items-center gap-4'>
      <Button size='small'>Button</Button>
      <Button>Button</Button>
      <Button size='large'>Button</Button>
    </div>
  ),
}

export const WithAnchor: Story = {
  args: {
    href: '#test',
    children: 'Button',
  },
  parameters: {
    docs: {
      description: 'Buttons with an `href` will be rendered as anchors',
    },
  },
}
