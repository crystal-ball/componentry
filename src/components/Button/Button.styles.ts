import { getMergedConfig } from '../../plugin-postcss/configs'

const { theme } = getMergedConfig()

//                                        <Button /> styles
// --------------------------------------------------------

export const Button = {
  // BASE
  '.ðŸ…²Button-base': {
    alignItems: 'center',
    display: 'inline-flex',
    justifyContent: 'center',
    lineHeight: 1, // ensures text is center aligned within flex layout

    userSelect: 'none', // Prevent text selection on click of buttons
    whiteSpace: 'nowrap', // By default button content shouldn't wrap

    '&.ðŸ…²-disabled': {
      cursor: 'default',
      pointerEvents: 'none',
    },
  },

  // VARIANTS
  '.ðŸ…²Button-filled': {
    height: '2rem',
    padding: '0 1rem',
    backgroundColor: theme.colors.primary[500],
    border: `1px solid ${theme.colors.primary[500]}`,
    borderRadius: theme.borderRadius.md,
    color: theme.colors.inverse,
    fontSize: theme.fontSize.small,

    transition: 'background-color 0.15s linear',

    '&:hover, &.ðŸ…²-hover': {
      borderColor: theme.colors.primary[500],
      backgroundColor: theme.colors.primary[700],
    },
    '&:active, &.ðŸ…²-active': {
      borderColor: theme.colors.primary[700],
      backgroundColor: theme.colors.primary[900],
    },
    '&.ðŸ…²-disabled': {
      borderColor: theme.colors.primary[300],
      backgroundColor: theme.colors.primary[300],
    },

    // For buttons with color options you'll typically define them in each variant
    // like:
    // '&.ðŸ…²Button-errorColor': { ... }
  },
  '.ðŸ…²Button-outlined': {
    height: '2rem',
    padding: '0 1rem',
    backgroundColor: 'transparent',
    border: `1px solid ${theme.colors.primary[500]}`,
    borderRadius: theme.borderRadius.md,
    color: theme.colors.primary[500],
    fontSize: theme.fontSize.small,

    transition: 'color 0.15s linear',

    '&:hover, &.ðŸ…²-hover': {
      borderColor: theme.colors.primary[700],
      color: theme.colors.primary[700],
    },
    '&:active, &.ðŸ…²-active': {
      borderColor: theme.colors.primary[900],
      color: theme.colors.primary[900],
    },
    '&.ðŸ…²-disabled': {
      borderColor: theme.colors.primary[300],
      color: theme.colors.primary[300],
    },
  },

  // SIZES
  '.ðŸ…²Button-smallSize': {
    height: '1.5rem',
    borderRadius: theme.borderRadius.md,
    fontSize: theme.fontSize.small,
    padding: '0rem 0.5rem',
  },
  '.ðŸ…²Button-largeSize': {
    height: '2.5rem',
    borderRadius: theme.borderRadius.lg,
    fontSize: theme.fontSize.large,
    padding: '0 2rem',
  },
}
