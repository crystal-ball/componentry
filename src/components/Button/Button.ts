import { useTheme } from '../Theme/Theme'
import { ComponentBaseProps } from '../../utils/types'
import { element } from '../../utils/element-creator'

interface ButtonProps extends ComponentBaseProps<'button'> {
  /** Button variant color */
  color?: 'primary'
  /** Disables the element, preventing mouse and keyboard events */
  disabled?: boolean
  /** Toggles full width element layout */
  fullWidth?: boolean
  /** HTML element href */
  href?: string
  /** Sets the display size */
  size?: 'sm' | 'lg'
  /** Display variant */
  variant?: 'filled' | 'outlined'
}

/**
 * [Button component 📝](https://componentry.design/components/button)
 * @experimental
 */
export const Button: React.FC<ButtonProps> = (props) => {
  const {
    variant = 'filled',
    color,
    fullWidth,
    size,
    ...merged
  } = {
    ...useTheme<ButtonProps>('Button'),
    ...props,
  }

  return element<ButtonProps>('Button', {
    // If an href is passed, this instance should render an anchor tag
    as: merged.href ? 'a' : 'button',
    type: merged.href ? undefined : 'button',
    componentCx: [
      `🅲Button-base 🅲Button-${variant}`,
      {
        [`🅲Button-${color}Color`]: color,
        [`🅲Button-${size}Size`]: size,
        'w-full': fullWidth,
      },
    ],
    ...merged,
  })
}
Button.displayName = 'Button'
