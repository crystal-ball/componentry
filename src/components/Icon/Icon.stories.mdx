import { ArgsTable, Meta, Story } from '@storybook/addon-docs'

import { Icon } from './Icon'

<Meta title='Components/Icon' component={Icon} />

# Icon

## Variants

Componentry ships with a single `font` variant which creates font icons:

<Story name='Icon'>
  <Icon id='coffee' />
</Story>

## Icon alignment

Icons are center aligned in flex and text contexts by default for convenience:

In flex containers icons are centered with `self-align: center`:

<h1 className='flex gap-0.5'>
  <Icon id='code' />
  Flex alignment
</h1>

<h2 className='flex gap-0.5'>
  <Icon id='code' />
  Flex alignment
</h2>

<h3 className='flex gap-0.5'>
  <Icon id='code' />
  Flex alignment
</h3>

In text elements icons are centered with `vertical-align: -0.15em`:

<h1>
  <Icon id='code' /> Flex alignment
</h1>

<h2>
  <Icon id='code' /> Flex alignment
</h2>

<h3>
  <Icon id='code' /> Flex alignment
</h3>

Checkmate... Hey, you know how I'm, like, always trying to save the planet? Here's my
chance <Icon id="code" matchTextBaseline />.They're using our own satellites against us.
And the clock is ticking. Must go faster... go, go, go, go, go! I was part of something
special. Remind me to thank John for a lovely weekend.

## Props

<ArgsTable of={Icon} />
