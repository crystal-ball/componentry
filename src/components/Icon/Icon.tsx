import React, { forwardRef } from 'react'
import { element } from '../../utils/element-creator'
import { MergeTypes, Resolve } from '../../utils/types'
import { UtilityProps } from '../../utils/utility-classes'
import { useThemeProps } from '../Provider/Provider'

// --------------------------------------------------------
// ICON ELEMENTS MAP

/** Mapping of icon IDs to components rendered by Icon */
export type IconElementsMap = { [ID: string]: React.ComponentType<unknown> }

let iconElementsMap: IconElementsMap = {}

/**
 * Configuration method for defining the elements to render for each Icon ID.
 * @remarks
 * Configuring an icon elements map isn't necessary if you've setup an SVG
 * symbol sprite.
 * @example
 * ```ts
 * import Info from './info.svg'
 * import Coffee from './coffee.svg'
 *
 * configureIconElementsMap({
 *   info: Info,
 *   coffee: Coffee
 * })
 * ```
 */
export function configureIconElementsMap(elementsMap: IconElementsMap): void {
  iconElementsMap = elementsMap
}

// --------------------------------------------------------
// ICON COMPONENT

/** Module augmentation interface for overriding component props' types */
export interface IconPropsOverrides {}

export interface IconPropsDefaults {
  /** External path to symbol sprite  */
  externalURI?: string
  /** ID for the `iconElementsMap` or href attribute for symbol sprites */
  id: string
  /** Display variant */
  variant?: 'font'
}

export type IconProps<Elem extends React.ElementType = 'svg'> = Resolve<
  MergeTypes<IconPropsDefaults, IconPropsOverrides>
> &
  UtilityProps &
  React.ComponentPropsWithRef<Elem> & { as?: Elem }

/**
 * Icon provides consistently themed iconography elements.
 * @example
 * ```tsx
 * <Icon id="coffee" />
 * ```
 * @see [üìù Icon](https://componentry.design/docs/components/icon)
 */
export interface Icon {
  <Elem extends React.ElementType = 'svg'>(
    props: IconProps<Elem>,
  ): React.ReactElement | null
  displayName?: string
}

export const Icon: Icon = forwardRef<HTMLElement, IconProps>((props, ref) => {
  const {
    externalURI = '',
    id,
    variant = 'font',
    ...rest
  } = { ...useThemeProps('Icon'), ...props }

  const hasMappedElement = id in iconElementsMap

  return element({
    ref,
    as: hasMappedElement ? iconElementsMap[id] : 'svg',
    children: hasMappedElement ? undefined : <use href={`${externalURI}#${id}`} />,
    componentCx: `C9Y-Icon-base C9Y-Icon-${variant} icon-${id}`,
    role: 'img',
    'aria-label': id,
    ...rest,
  })
})
Icon.displayName = 'Icon'
