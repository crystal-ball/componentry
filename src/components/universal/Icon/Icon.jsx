// @flow
import React, { Component } from 'react'
import classNames from 'classnames'

type Props = {
  className: ?string,
  font: ?boolean,
  id: string,
  title: ?string,
  presentation: ?boolean,
}

/**
 * Use the icon component with an svg imported from `/media/icons` to create an icon
 * font.
 */
export default class Icon extends Component<Props> {
  static defaultProps = {
    className: '',
    font: true,
    presentation: false,
    title: null,
  }

  constructor(props: Props) {
    super(props)
    // $FlowIgnore: Flow doesn't know about webpack magic comments for dynamic imports
    import(/* webpackMode: "eager", webpackChunkName: "icons" */ `media/icons/${
      props.id
    }.svg`)
  }

  render() {
    const { title, presentation, id, className, font } = this.props

    return (
      <svg
        aria-label={title}
        role={presentation ? 'presentation' : 'img'}
        className={classNames('icon', id, className, { font })}
      >
        <use xlinkHref={`#${id}`} href={`#${id}`} />
      </svg>
    )
  }
}
