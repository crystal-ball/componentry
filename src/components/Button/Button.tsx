import clsx from 'clsx'
import { type ForwardedRef, type ReactElement, type ReactNode, forwardRef } from 'react'
import { type ComponentBaseProps } from '../../utils/base-types'
import { element } from '../../utils/element-creator'
import { type MergePropTypes } from '../../utils/types'
import { Icon } from '../Icon/Icon'
import { useThemeProps } from '../Provider/Provider'

// Module augmentation interface for overriding component props' types
export interface ButtonPropsOverrides {}

export interface ButtonPropsDefaults {
  children: ReactNode
  /** Button variant color */
  color?: 'primary'
  /** Disables the element, preventing mouse and keyboard events */
  disabled?: boolean
  /** Icon positioned after button content */
  endIcon?: string | JSX.Element
  /** Toggles full width element layout */
  fullWidth?: boolean
  /** HTML element href */
  href?: string
  /** Sets the display size */
  size?: 'small' | 'large'
  /** Icon positioned before button content */
  startIcon?: string | JSX.Element
  /** Display variant */
  variant?: 'filled' | 'outlined'
  /** Indicates whether buttons in a disabled state should be wrapped with a span */
  wrapWhenDisabled?: boolean
}

export type ButtonProps = MergePropTypes<ButtonPropsDefaults, ButtonPropsOverrides> &
  ComponentBaseProps<'button'>

// ‚ú® Nice display type for IntelliSense
export interface Button {
  (props: ButtonProps & { ref?: ForwardedRef<unknown> }): ReactElement | null
  displayName?: string
}

/**
 * #### [üìù Button](https://componentry.design/docs/components/button)
 *
 * Button provides actionable elements for creating accessible user interactions.
 * @example
 * ```tsx
 * <Button onClick={() => buildSomethingRadical()}>
 *   Build something radical
 * </Button>
 * ```
 */
export const Button: Button = forwardRef((props, ref) => {
  const {
    variant = 'filled',
    children,
    disabled,
    startIcon,
    endIcon,
    color,
    fullWidth,
    size,
    wrapWhenDisabled = true,
    ...merged
  } = {
    ...useThemeProps('Button'),
    ...props,
  }

  const iconCx = clsx('C9Y-Button-Icon', { [`C9Y-Button-Icon-${size}Size`]: size })
  const decoratedStartIcon =
    startIcon === undefined ? null : typeof startIcon === 'string' ? (
      <Icon id={startIcon} className={iconCx} />
    ) : (
      <span className={iconCx}>{startIcon}</span>
    )

  const decoratedEndIcon =
    endIcon === undefined ? null : typeof endIcon === 'string' ? (
      <Icon id={endIcon} className={iconCx} />
    ) : (
      <span className={iconCx}>{endIcon}</span>
    )

  const contents = element({
    ref,
    disabled,
    // If an href is passed, this instance should render an anchor tag
    as: merged.href ? 'a' : 'button',
    type: merged.href ? undefined : 'button',
    componentCx: [
      `C9Y-Button-base C9Y-Button-${variant}`,
      {
        [`C9Y-Button-${color}Color`]: color,
        [`C9Y-Button-${size}Size`]: size,
        'w-full': fullWidth,
      },
    ],
    children: (
      <>
        {decoratedStartIcon}
        {children}
        {decoratedEndIcon}
      </>
    ),
    ...merged,
  })

  return disabled && wrapWhenDisabled ? (
    <span className={clsx('C9Y-Button-DisabledWrapper', { 'w-full': fullWidth })}>
      {contents}
    </span>
  ) : (
    contents
  )
})
Button.displayName = 'Button'
