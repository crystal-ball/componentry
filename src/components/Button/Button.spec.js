import { render, screen } from '@testing-library/react'

import { elementTests } from '../../test/element-tests'
import { Icon } from '../Icon/Icon'
import { Button } from './Button'

describe('<Button/>', () => {
  // Basic library element test suite
  elementTests(Button)

  // 3. Test default as is 'button'
  // 4. Test passing block provides $variant-block
  // 5. Test passing color provides $variant-color
  // 6. Test passing outline provides $variant-outline-$outline
  // 7. Test passing size provides $variant-size
  // 8. Test passing disabled

  it('when no props are passed, then defaults should be rendered', () => {
    render(<Button>Button</Button>)

    // By default the button should have type button for a11y
    expect(screen.getByRole('button')).toHaveAttribute('type', 'button')
    // By default the variant primary
    expect(screen.getByRole('button')).toHaveClass('C9Y-Button-base C9Y-Button-filled')
    // TODO: Is it possible to get the classes from the element and just check that?
    // ...there shouldn't be any other classes
  })

  it('when `type` is passed, then it overrides the default', () => {
    render(<Button type='submit'>Button</Button>)

    expect(screen.getByRole('button')).toHaveAttribute('type', 'submit')
  })

  it('when `variant` is passed, then it should be used as base className value', () => {
    render(<Button variant='outlined'>Button</Button>)

    expect(screen.getByRole('button')).toHaveClass('C9Y-Button-base C9Y-Button-outlined')
  })

  it('when `color` is passed, then the color className should render', () => {
    render(
      <>
        <Button color='primary'>Button</Button>
        <Button variant='outlined' color='primary'>
          Outlined Button
        </Button>
      </>,
    )

    expect(screen.getByRole('button', { name: 'Button' })).toHaveClass(
      'C9Y-Button-base C9Y-Button-filled C9Y-Button-primaryColor',
    )
    expect(screen.getByRole('button', { name: 'Outlined Button' })).toHaveClass(
      'C9Y-Button-base C9Y-Button-outlined C9Y-Button-primaryColor',
    )
  })

  it('when `size` is passed, then the size className should render', () => {
    render(
      <>
        <Button size='small'>Button</Button>
        <Button variant='outlined' size='small'>
          Outlined Button
        </Button>
      </>,
    )

    expect(screen.getByRole('button', { name: 'Button' })).toHaveClass(
      'C9Y-Button-base C9Y-Button-filled C9Y-Button-smallSize',
    )
    expect(screen.getByRole('button', { name: 'Outlined Button' })).toHaveClass(
      'C9Y-Button-base C9Y-Button-outlined C9Y-Button-smallSize',
    )
  })
})

describe('<Button /> - Icons', () => {
  it('renders React elements passed as buttons', () => {
    render(
      <Button
        startIcon={<Icon id='code' data-special='very' data-testid='custom-one' />}
        endIcon={<Icon id='code' data-awesome='much' data-testid='custom-two' />}
      >
        Componentry
      </Button>,
    )

    expect(screen.getByTestId('custom-one').dataset.special).toBe('very')
    expect(screen.getByTestId('custom-two').dataset.awesome).toBe('much')
  })

  it('renders Icon components for string values', () => {
    render(
      <Button startIcon='code' endIcon='code'>
        Componentry
      </Button>,
    )

    expect(screen.getAllByLabelText('code')).toHaveLength(2)
  })
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Button /> - Snapshots', () => {
  it('renders defaults correctly', () => {
    render(<Button>Componentry</Button>)

    expect(screen.getByRole('button')).toMatchSnapshot()
  })
})
