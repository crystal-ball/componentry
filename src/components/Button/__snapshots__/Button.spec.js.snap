// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Snapshots renders brand color correctly 1`] = `
<button
  class="ðŸ…²Button-base ðŸ…²Button-filled ðŸ…²Button-successColor"
  type="button"
>
  Componentry
</button>
`;

exports[`<Button /> Snapshots renders defaults correctly 1`] = `
<button
  class="ðŸ…²Button-base ðŸ…²Button-filled"
  type="button"
>
  Componentry
</button>
`;

exports[`<Button /> Snapshots renders large outline correctly 1`] = `
<button
  class="ðŸ…²Button-base ðŸ…²Button-filled ðŸ…²Button-lgSize"
  type="button"
>
  Componentry
</button>
`;
